### PR TITLE
Implemented Cart and List voice commands, also a String Extension met…

### PR DESCRIPTION
**Pull request by marjasi**
…hod.

### CODE REVIEW

**Comment by marjasi** [comment]
[[Voice] Voice status (like shopping cart, shopping list...)](https://trello.com/c/mK00jj0J/86-voice-voice-status-like-shopping-cart-shopping-list)

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 1]
```diff
@@ -1,50 +1,171 @@
...
+            }
+        }
+
+        private void SpeakOut(string message)
+        {
+            if (!string.IsNullOrEmpty(message))
+            {
+                tts.Speak(message, QueueMode.Flush, null, null);
+
+                while (tts.IsSpeaking) { }
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs]: We need to wait until the tts has finished speaking. Don't know how to write this differently.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 2]
```diff
@@ -1,50 +1,171 @@
...
+                Regex addProductRegex = new Regex(@"^" + @ConfigurationManager.AppSettings["CmdAddCartList"] + @"(?:\s\w+)+");
+
+                if (matches[0] == voiceCartCmd)
+                {
+                    string endMessage = "Voicing of cart complete.";
+
+                    foreach (string item in prefs.GetPreferencesToList())
+                    {
+                        SpeakOut(item);
+                        Thread.Sleep(int.Parse(ConfigurationManager.AppSettings["VoicerPauseTime"]));
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs]: We need to pause between each item so that the user has time to contemplate the results.

File: Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace ShopLens.Extensions
+{
+    static class ExtensionMethods
+    {
+        public static string FirstCharToUpper(this string input)
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs:10]: Finally! :+1: :+1: :+1:

File: Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            switch (input)
+            {
+                case null:
+                    throw new ArgumentNullException(nameof(input));
+
+                case "":
+                    throw new ArgumentException($"{nameof(input)} cannot be empty", nameof(input));
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs]: Don't think this should throw if you ask me, just return empty string.

File: Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+                default:
+                    return input.First().ToString().ToUpper() + input.Substring(1);
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs:21]: By the way, this class is just begging :pray: to be tested :heavy_check_mark: / :x: 

Like here, I'm not sure if `.Substring(1)` will throw an exception if the string is just 1 letter?

(I think that the docs say it won't throw in this case actually https://docs.microsoft.com/en-us/dotnet/api/system.string.substring?view=netframework-4.7.2)

Again, just want you to start spotting these places where unit tests are great :eyes:

File: Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿namespace ShopLens.Droid.Helpers
+{
+    public static class ProductNameHelper
+    {
+        public static string GetProductNameFromString(string products)
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs]: Also begging to be tested :pray:

File: Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                    if (i == 1)
+                    {
+                        productName += array[i];
+                    }
+                    else
+                    {
+                        productName += " " + array[i];
+                    }
+                }
+            }
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs]: First impression: what in the world is this loop :astonished: 

Again, if this was unit tested, I'd go to the unit test and look what input it expects and what output it produces. (Not saying you should go write those tests right now)

Ok, so you're skipping first element in the list for some reason. Perhaps it was the "add" word or something (why is this method responsible for that?). Anyway, in that case I would write instead of 

```
if (i > 0)
{
// blabla
}
```

Something like this:

```
if (i == 0) break;
//blabla
```

How is this better? Firstly, it reads better - "if it's the first element, skip it".

Secondly, it avoids nesting "if"s. [Link](https://www.reddit.com/r/csharp/comments/33puzj/is_it_a_good_or_bad_practice_reducing_nested_if/)

To sum up, I think this method should be rewritten. Either the first word (the verb "add"/"delete"?) should not be passed to the method, or the amount of words to skip should be passed as a parameter. (I think first is better, as it is more Solid - single responsibility).

The `for` loop should be rewritten into a `foreach`, if possible. Use LINQ methods like `.Skip(1)`, if absolutely needed, as they are more readable. Actually, with this much logic, I think the whole method can be rewritten as:

```
productName = String.Join(array.Skip(1), " ")     
```

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs (hunk 2 above)
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs]: So, every time the sleep is called, the configuration manager will open a file, read it whole, find the value of "VoicePauseTime", and parse it? :thinking: 

Do you get what I'm saying?

**Review by EdvardasDlugauskas — CHANGES_REQUESTED** [review]
This good stuff is actually mergable :100: :ok_hand: 

However, if you have time, take a look at the comments. :+1:

File: Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs (hunk 4 above)
Comment (marjasi) [reply Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs]: Well, I didn't write this method, but I thought I should leave the throws.

File: Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs (hunk 5 above)
Comment (marjasi) [reply Mobile/ShopLens/ShopLens/Extensions/ExtensionMethods.cs:21]: Ok, so I should write a Unit test? Again, not my code here.

File: Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs (hunk 7 above)
Comment (marjasi) [reply Mobile/ShopLens/Droid/Helpers/ProductNameHelper.cs]: Oh ok, so I shouldn't write unit tests, gotcha. While writing this method I thought I was writing something I'd have written back during the first semester. I guess that was a warning sign to not write this stuff...

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs (hunk 2 above)
Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs]: Does it actually parse the file everytime, though? It might create a local dictionary from the file and use that instead. But I don't really know.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 8]
```diff
@@ -147,12 +147,24 @@ public void OnResults(Bundle results)
...
+                    string endMessage = itemToAdd + " was added to your shopping cart.";
 
                     AddStringToList(itemToAdd);
 
-                    SpeakOut(endMessage);
+                    Task.Run(() =>
+                    {
+                        SpeakOut(endMessage);
+
+                    }).ContinueWith((t) =>
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:160]: Good stuff 👍

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs [hunk 9]
```diff
@@ -106,24 +111,36 @@ public void OnResults(Bundle results)
             var matches = results.GetStringArrayList(SpeechRecognizer.ResultsRecognition);
             if (matches.Count > 0)
             {
-                Regex addProductRegex = new Regex(@"^" + ConfigurationManager.AppSettings["CmdAddCartList"] + @"(?:\s\w+)+");
+                string cmdAddProduct = ConfigurationManager.AppSettings["CmdAddCartList"];
+                Regex addProductRegex = new Regex(@"^" + cmdAddProduct + @"(?:\s\w+)+");
 
                 if (matches[0] == voiceListCmd)
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs]: `matches[0]` is used like a bazillion times in this method, why not just assign it to a variable with a meaningful name to improve readability? It has better performance too

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs [hunk 10]
```diff
@@ -106,24 +111,36 @@ public void OnResults(Bundle results)
...
+                        {
+                            System.Diagnostics.Debug.WriteLine(t.Exception);
+                        }
+                    }); 
                 }
 
                 if (addProductRegex.IsMatch(matches[0]))
                 {
-                    string itemToAdd = ProductNameHelper.GetProductNameFromString(matches[0]).FirstCharToUpper();
+                    string itemToAdd = string.Join("", matches[0].Skip(cmdAddProduct.Length + 1)).FirstCharToUpper();
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs]: "👆 Va čia pagalvokite labai gerai, kodėl čia negerai, pagalvokite... " 🏃

Hint: why do you need the `Join`? If you make matches[0] a variable maybe it will be more clear

**Review by EdvardasDlugauskas — APPROVED** [review]

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs (hunk 10 above)
Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs]: That running man emote looked so different on my mobile phone.

Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs]: Where did you get that quote from? Seems too cliche to be original.
